### ISSUE TITLE
CORS is not permitted between multiple subdomain workspace apps 

### ISSUE THREAD

**Issue by palmmymumu** [coder/coder#5706]
I'm using coder with caddy to dev my backend project but when using forwarded port api url with XMLHttpRequest (Vue axios) it occurs error

`Access to XMLHttpRequest at 'https://8001--dev--xxx--xxx.coder.xxx.com/admin/report?start=2023-01-13&end=2023-01-13' from origin 'https://3000--dev--xxx--xxx.coder.xxx.com' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: Redirect is not allowed for a preflight request.`

I saw in chrome dev tools when frontend trying to get preflight request (options request) to that path, it returned http code 307 with redirection to **"https://coder.xxx.com/api/v2/applications/auth-redirect?redirect_uri=https%3A%2F%2F8001--xxx--xxx--xxx.coder.xxx.com%2Fadmin%2Freport%3Fstart%3D2023-01-13%26end%3D2023-01-13"**

Which CORS policy doesn't allow. What should I do?

**Comment by kylecarbs** [comment]
Ahh, interesting. I believe we need to adjust our CORS policy. cc @deansheather

**Comment by deansheather** [comment]
I believe we don't allow any requests to other app subdomains for security purposes, so even if we didn't redirect we would still block this request anyways. The reason is that this could allow you to connect to and make requests to other app URLs from the browser. e.g. you could make a malicious javascript page that makes requests to an app on someone's workspace.

We should implement CORS properly for apps though, I believe the app handler is above the CORS handler in the middleware stack so we're not sending CORS properly. Or we don't have CORS anywhere yet.

@palmmymumu I believe you could fix the problem by mounting the 8001 and 3000 ports on different paths of your webserver. For example, if the two apps are an API backend and a frontend server, you could mount the frontend at `/` and the backend at `/api`. Since you use caddy already this should be a breeze.

**Comment by pietromarchesani** [comment]
I'am having the exact same problem in a multi workspaces scenario, where all workspaces are owned by the same user. It is done to recreate the exact same scenario that would run in a production environment. Let's say you can have 1 workspace for frontend development (i.e., react) and 2 workspaces for backend development (spring boot microservices). Such scenario would not work. I am forces to proxy requests at frontend side, otherwise I would run in CORS issue. How this scenario can be solved in more efficient way?

**Comment by pietromarchesani** [comment]
@deansheather, If I got your point correctly this would allow CORS within the same user workspace only - a single one - which would not improve cloud native app development that much.
On top of that, forcing it thru app pre-defined in the user workspace would not leave much room for flexibility. I guess the workspace needs to be build (from template) having already in mind the app is going to be developed.
In a more complex scenario a user might develop both frontend and backend services which are spreading in multiple workspaces in order to improve cloud native development experience.
It would be great improving DX at such level.
General workspace access is already protected by authentication.
In addition on that CORS could be allowed using "Wildcard access URL", thus services might be accessed by workspaces running on the same Coder instance, i.e. probably same company. It could be another level of "protection".
Finally, a feature flag could be added at Coder instance, i.e. ALLOW_CORS_FOR_WILDCARD_DOMAIN, disabled by default, so any client can decide to enable it at its own risk.

**Comment by elliot-graebert-skydio** [comment]
Adding a +1 here for needing a solution. The specific use case I have is:

- On 1 workspace, I have 2 processing running: react app on port 3000 and backend server on 4000
- I'm trying to access these ports via forwarded urls.
- When the backend server is redirecting requests, it triggers the CORS issue in Chrome.

The consequence is that it's not possible to develop my app using port forwarded URLs. This means that everyone here that is trying to use Coder has to install the CLI locally and set up port forwarding. If this one issue was fixed, people would be able to operate without making any local changes.

I don't know enough about CORS to give an official suggestion on what would work here, though from chatting with @deansheather , I think he's got a good path forward.

**Comment by junglesub** [comment]
@elliot-graebert-skydio I ran into the same issue :(

Also did some testing and realized that I only get cors preflight error (client → backend) when header is included.  
(Such as `"Content-Type": "application/json"`)

I also think that it removes the body data but I am not sure :/

Any thoughts?

**Comment by code-asher** [comment]
Reopening this because although #7688 sets the proper CORS headers if you do not authenticate to the other app beforehand the request will fail due to an auth redirect so we need to figure out a solution for that.

**Comment by deansheather** [comment]
This could be done by adding a check to the app serving code for OPTIONS requests that either returns access control headers that block the request or headers that allow from `*.$app_host` depending on whether the request is originating from an app in the same workspace.

I.e. if your app on `app1--agent1--workspace1--user.site.com` wants to connect to `app2--agent2--workspace1--user.site.com` this should be allowed as they are from the same workspace. However, `app1--agent1--workspace1--user.site.com` should not be able to connect to `app3--agent3--workspace2--user.site.com` or any workspaces owned by a different user.

This should be a dangerous site flag and disabled by default

**Comment by deansheather** [comment]
Investigation is needed into whether it's reliable to conditionally serve access control headers based on referrer/origin/host headers. I think it should work but not fully sure

**Comment by code-asher** [comment]
Yup I think the problem now is only non-OPTIONS requests.  Since we set a cookie on each subdomain when there are not yet any cookies for the other subdomain to send the request gets redirected to auth.

I think we would have to set CORS headers on the auth endpoints as well or set the cookies in some other way.

**Comment by code-asher** [comment]
Actually I am not sure this is actually a problem.  I logged out, accessed an app on workspace 1, logged in, made a request from app 1 to app 2, and it worked fine.  It only breaks when I explicitly go to app 2 and delete my cookies.

**Comment by code-asher** [comment]
OK yeah looks like we actually set two cookies, one on the subdomain and another on the root.  As long as the root one exists I can make requests from any other app and the subdomain cookie gets set successfully.  Since you have to log into the one app in the first place that root cookie will always be set which I think means there is nothing more we need to do here.  @Emyrk lemme know if I missed something.

**Comment by bpmct** [comment]
For anyone searching, you can view these docs on our CORS behavior: https://coder.com/docs/v2/latest/networking/port-forwarding#cross-origin-resource-sharing-cors

**Comment by szab100** [comment]
We are still having problems with this, the behavior we are experiencing does NOT seem to match the documentation's table. The use-case is the following:

We have a React app running in the workspace on port 34212, with port forward url `https://34212--main--<workspace-name>--<username>.<coder-access-url>/` that tries to make requests to the backend running within the same workspace / agent on port 8080 with port forward url `https://8080--main--<workspace-name>--<username>.<coder-access-url>/`. As per my understanding, based on the [documentation](https://coder.com/docs/v2/latest/networking/port-forwarding#cross-origin-resource-sharing-cors), this should be allowed, even without `credentials: "include"`, which we set in the fetch request anyway.

But requests fail with the following error in Chrome console:
`Access to fetch at 'https://8080--main--<workspace-name>--<username>.<coder-access-url>/v1/xyz' from origin 'https://34212--main--<workspace-name>--<username>.<coder-access-url>' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`

The OPTIONS request copied as curl (exact workspace name, username, access url redacted): 
`curl 'https://8080--main--<workspace-name>--<username>.<coder-access-url>/v1/xyz' \
  -X 'OPTIONS' \
  -H 'authority: 8080--main--<workspace-name>--<username>.<coder-access-url>' \
  -H 'accept: */*' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'access-control-request-headers: authorization,content-type,cws_auth_token,intuit_tid' \
  -H 'access-control-request-method: GET' \
  -H 'origin: https://34212--main--<workspace-name>--<username>.<coder-access-url>' \
  -H 'referer: https://34212--main--<workspace-name>--<username>.<coder-access-url>/' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-site' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36' \
  --compressed`

Response headers:
`Content-Length: 0
Date: Thu, 10 Aug 2023 00:48:11 GMT
Vary: Access-Control-Request-Headers
Vary: Access-Control-Request-Method
Vary: Origin`

**Comment by szab100** [comment]
@bpmct @deansheather We are still having problems with this, the behavior we are experiencing does **not** seem to match the documentation's table. The use-case is the following:

We have a React app running in the workspace on port 34212, with port forward url `https://34212--<agent-name>--<workspace-name>--<username>.<coder-access-url>/` that tries to make requests to the backend running within the same workspace / agent on port 8080 with port forward url `https://8080--<agent-name>--<workspace-name>--<username>.<coder-access-url>/`. As per my understanding, based on the [documentation](https://coder.com/docs/v2/latest/networking/port-forwarding#cross-origin-resource-sharing-cors), this should be allowed, even without `credentials: "include"`, which we set in the fetch request anyway. The React web UI on the port-forward URL is accessed by the workspace owner, so it's the most basic scenario, no cross-workspace / cross-user access involved.

But when the React UI loads, all the requests it tries to make to the backend using the 8080 port forward URL fail with the following CORS error in Chrome console:
```Access to fetch at 'https://8080--<agent-name>--<workspace-name>--<username>.<coder-access-url>/v1/xyz' from origin 'https://34212--<agent-name>--<workspace-name>--<username>.<coder-access-url>' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.```

The problem is likely that the `coder_devurl_signed_app_token` JWT Cookie's domain is set to exactly match the 34212 port-fwd URL, so regardless of `credentials: "include"`, the browser doesn't send it along with the requests, as you can see below 😢

Luckily, since we have a reverse proxy between the Coder deployment & user requests, we can easily pattern match port-fwd URLs & override the returned CORS headers as a quick & dirty workaround (set allow origins to the request's Origin value). But with this, we effectively disable Coder's CORS protections. 

To properly fix this on Coder side, I believe the port forward URL scheme has to be changed to something like `<port>.<agent-name>--<workspace-name>--<username>.<coder-access-url>`, eg the first field separator changed to dot instead of double-dashes, so the JWT Cookie's domain can be set to `.<agent-name>--<workspace-name>--<username>.<coder-access-url>` and therefore the browser will send the JWT along with the Fetch requests to all port-forward URLs within the same workspace / agent and the behavior described in the docs can be achieved.

The OPTIONS request copied as curl: 
```curl 'https://8080--<agent-name>--<workspace-name>--<username>.<coder-access-url>/v1/xyz' \
  -X 'OPTIONS' \
  -H 'authority: 8080--<agent-name>--<workspace-name>--<username>.<coder-access-url>' \
  -H 'accept: */*' \
  -H 'accept-language: en-US,en;q=0.9' \
  -H 'access-control-request-headers: authorization,content-type,cws_auth_token,intuit_tid' \
  -H 'access-control-request-method: GET' \
  -H 'origin: https://34212--<agent-name>--<workspace-name>--<username>.<coder-access-url>' \
  -H 'referer: https://34212--<agent-name>--<workspace-name>--<username>.<coder-access-url>/' \
  -H 'sec-fetch-dest: empty' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-site: same-site' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36' \
  --compressed```

Response headers:
```Content-Length: 0
Date: Thu, 10 Aug 2023 00:48:11 GMT
Vary: Access-Control-Request-Headers
Vary: Access-Control-Request-Method
Vary: Origin```

**Comment by szab100** [comment]
Ooops, I just realized that we are still on a fairly old Coder version (0.23.7), just 1 minor version before the fix was released in 0.24.0! 🤦 🤦 🤦  

So please disregard my entire post above, will test this again once we did with the version upgrade and post again if there are still any outstanding issues. But likely won't be, after looking at the diff in the merged PR (#7688). 👍
